### ISSUE TITLE
Backup/Restore Table with on cluster Clause Fails when use Azure Blob Storage as Backup Disk

### ISSUE THREAD

**Issue by JeremyChen-179** [ClickHouse/ClickHouse#52506]
Encountering issues when trying to backup and restore tables using the `on cluster` clause with Azure Blob Storage as the backup disk.

## **Environment:**

ClickHouse server version: 23.6.2
Cluster: 3 replicas, all three nodes mounted with the same Azure Blob storage
Disk Configuration: blob_storage_disk on all three nodes
```sh
<storage_configuration>
    <disks>
        <blob_storage_disk>
            <type>azure_blob_storage</type>
            <storage_account_url></storage_account_url>
            <container_name>backup</container_name>
            <account_name></account_name>
            <account_key></account_key>
            <metadata_path>/var/lib/clickhouse/disks/blob_storage_disk/</metadata_path>
            <cache_enabled>true</cache_enabled>
            <cache_path>/var/lib/clickhouse/disks/blob_storage_disk/cache/</cache_path>
            <skip_access_check>false</skip_access_check>
        </blob_storage_disk>
    </disks>
    </storage_configuration>
    <backups>
        <allowed_disk>blob_storage_disk</allowed_disk>
    </backups>
```

### Backup Issue:

1. Execute BACKUP TABLE test ON CLUSTER cluster_1S_3R TO Disk('blob_storage_disk', 'test_on_cluster-21-07-23') on cluster_1S_3R node 2
2. The above command fails with the error:
```sh
Received exception from server (version 23.6.2):
Code: 655. DB::Exception: Received from localhost:9000. DB::Exception: Got error from clickhouse%2D03:9000. DB::Exception: Lock file .lock suddenly disappeared while writing backup Disk('blob_storage_disk', 'test_on_cluster-21-07-23'). (FAILED_TO_SYNC_BACKUP_OR_RESTORE)
```

However, running a backup without the `on cluster` clause is successful. 
```sh
cluster_1S_3R node 2 :) backup table test TO Disk('blob_storage_disk', 'test-21-07-23')

BACKUP TABLE test TO Disk('blob_storage_disk', 'test-21-07-23')

Query id: 53d38130-4500-4722-81a1-456a1755598d

┌─id───────────────────────────────────┬─status─────────┐
│ 205e3b8e-5f6f-4280-920c-4efadb5416fb │ BACKUP_CREATED │
└──────────────────────────────────────┴────────────────┘

1 row in set. Elapsed: 8.934 sec.
```
And the backup file is created in Azure Blob Storage with a random name

### Restore Issue:

1. Execute RESTORE TABLE test AS test_restored ON CLUSTER cluster_1S_3R FROM Disk('blob_storage_disk', 'test-21-07-23') on cluster_1S_3R node 2
2. This command fails with the error:
```sh
Received exception from server (version 23.6.2):
Code: 599. DB::Exception: Received from localhost:9000. DB::Exception: Got error from clickhouse%2D03:9000. DB::Exception: Backup Disk('blob_storage_disk', 'test-21-07-23') not found. (BACKUP_NOT_FOUND)
```

The error suggests that the backup cannot be found, even though the backup exists in Azure Blob Storage (verified by the successful backup operation without on cluster). It's noteworthy to mention that, restore without on cluster is not supported for RMT tables.

The detailed trace logs are included in the report. It would be great if you can investigate this issue and provide guidance on how to resolve it. Please let me know if you need any further information.

```sh
RESTORE TABLE test AS test_restored ON CLUSTER cluster_1S_3R FROM Disk('blob_storage_disk', 'test-21-07-23')

Query id: 3894ca6e-0aa6-40ec-ac36-65359d2b95c5

[clickhouse-02] 2023.07.24 02:59:03.835746 [ 41 ] {3894ca6e-0aa6-40ec-ac36-65359d2b95c5} <Debug> executeQuery: Query span trace_id for opentelemetry log: 00000000-0000-0000-0000-000000000000
[clickhouse-02] 2023.07.24 02:59:03.836074 [ 41 ] {3894ca6e-0aa6-40ec-ac36-65359d2b95c5} <Debug> executeQuery: (from 127.0.0.1:58166) RESTORE TABLE test AS test_restored ON CLUSTER cluster_1S_3R FROM Disk('blob_storage_disk', 'test-21-07-23') (stage: Complete)
[clickhouse-02] 2023.07.24 02:59:03.885186 [ 41 ] {3894ca6e-0aa6-40ec-ac36-65359d2b95c5} <Trace> RestorerFromBackup: Will use paths in backup: "/"
[clickhouse-02] 2023.07.24 02:59:03.885197 [ 41 ] {3894ca6e-0aa6-40ec-ac36-65359d2b95c5} <Trace> RestorerFromBackup: Finding tables in backup
[clickhouse-02] 2023.07.24 02:59:03.909556 [ 41 ] {3894ca6e-0aa6-40ec-ac36-65359d2b95c5} <Information> RestorerFromBackup: Will restore 0 databases and 1 tables
[clickhouse-02] 2023.07.24 02:59:03.909905 [ 41 ] {3894ca6e-0aa6-40ec-ac36-65359d2b95c5} <Trace> ContextAccess (default): Access granted: INSERT, CREATE TABLE ON default.test_restored
[clickhouse-02] 2023.07.24 02:59:03.909936 [ 41 ] {3894ca6e-0aa6-40ec-ac36-65359d2b95c5} <Trace> RestorerFromBackup: Will use paths in backup: "/"
[clickhouse-02] 2023.07.24 02:59:03.909939 [ 41 ] {3894ca6e-0aa6-40ec-ac36-65359d2b95c5} <Trace> RestorerFromBackup: Finding tables in backup
[clickhouse-02] 2023.07.24 02:59:03.933558 [ 41 ] {3894ca6e-0aa6-40ec-ac36-65359d2b95c5} <Information> RestorerFromBackup: Will restore 0 databases and 1 tables
[clickhouse-02] 2023.07.24 02:59:03.933575 [ 41 ] {3894ca6e-0aa6-40ec-ac36-65359d2b95c5} <Trace> ContextAccess (default): Access granted: INSERT, CREATE TABLE ON default.test_restored
[clickhouse-02] 2023.07.24 02:59:03.933594 [ 41 ] {3894ca6e-0aa6-40ec-ac36-65359d2b95c5} <Trace> RestorerFromBackup: Will use paths in backup: "/"
[clickhouse-02] 2023.07.24 02:59:03.933607 [ 41 ] {3894ca6e-0aa6-40ec-ac36-65359d2b95c5} <Trace> RestorerFromBackup: Finding tables in backup
[clickhouse-02] 2023.07.24 02:59:03.957387 [ 41 ] {3894ca6e-0aa6-40ec-ac36-65359d2b95c5} <Information> RestorerFromBackup: Will restore 0 databases and 1 tables
[clickhouse-02] 2023.07.24 02:59:03.957411 [ 41 ] {3894ca6e-0aa6-40ec-ac36-65359d2b95c5} <Trace> ContextAccess (default): Access granted: INSERT, CREATE TABLE ON default.test_restored
[clickhouse-02] 2023.07.24 02:59:03.966182 [ 41 ] {3894ca6e-0aa6-40ec-ac36-65359d2b95c5} <Information> RestoreCoordinationRemote: Waiting for the stage completed
[clickhouse-02] 2023.07.24 02:59:03.967195 [ 41 ] {3894ca6e-0aa6-40ec-ac36-65359d2b95c5} <Trace> RestoreCoordinationRemote: Seems like host (clickhouse%2D01:9000) is dead. Will retry the check to confirm
[clickhouse-02] 2023.07.24 02:59:03.967386 [ 41 ] {3894ca6e-0aa6-40ec-ac36-65359d2b95c5} <Trace> RestoreCoordinationRemote: ZooKeeperRetriesControl: RestoreCoordinationRemote/readCurrentState::checkAliveNode: setKeeperError: error=ZNONODE message=There is no alive node for host clickhouse%2D01:9000. Will retry
[clickhouse-02] 2023.07.24 02:59:03.967395 [ 41 ] {3894ca6e-0aa6-40ec-ac36-65359d2b95c5} <Debug> RestoreCoordinationRemote: ZooKeeperRetriesControl: RestoreCoordinationRemote/readCurrentState::checkAliveNode: will retry due to error: retry_count=1 timeout=100ms error=ZNONODE message=There is no alive node for host clickhouse%2D01:9000. Will retry
[clickhouse-02] 2023.07.24 02:59:04.069215 [ 41 ] {3894ca6e-0aa6-40ec-ac36-65359d2b95c5} <Debug> RestoreCoordinationRemote: ZooKeeperRetriesControl: RestoreCoordinationRemote/readCurrentState::checkAliveNode: succeeded after: iterations=2 total_retries=1
[clickhouse-02] 2023.07.24 02:59:04.069240 [ 41 ] {3894ca6e-0aa6-40ec-ac36-65359d2b95c5} <Trace> RestoreCoordinationRemote: Host (clickhouse%2D01:9000) appeared to be alive
[clickhouse-02] 2023.07.24 02:59:04.069245 [ 41 ] {3894ca6e-0aa6-40ec-ac36-65359d2b95c5} <Trace> RestoreCoordinationRemote: Seems like host (clickhouse%2D02:9000) is dead. Will retry the check to confirm
[clickhouse-02] 2023.07.24 02:59:04.069444 [ 41 ] {3894ca6e-0aa6-40ec-ac36-65359d2b95c5} <Trace> RestoreCoordinationRemote: Host (clickhouse%2D02:9000) appeared to be alive
[clickhouse-02] 2023.07.24 02:59:04.069450 [ 41 ] {3894ca6e-0aa6-40ec-ac36-65359d2b95c5} <Trace> RestoreCoordinationRemote: Seems like host (clickhouse%2D03:9000) is dead. Will retry the check to confirm
[clickhouse-02] 2023.07.24 02:59:04.069621 [ 41 ] {3894ca6e-0aa6-40ec-ac36-65359d2b95c5} <Trace> RestoreCoordinationRemote: Host (clickhouse%2D03:9000) appeared to be alive
[clickhouse-02] 2023.07.24 02:59:04.069630 [ 41 ] {3894ca6e-0aa6-40ec-ac36-65359d2b95c5} <Information> RestoreCoordinationRemote: Waiting on ZooKeeper watch for any node to be changed (currently waiting for host clickhouse%2D01:9000)
[clickhouse-02] 2023.07.24 02:59:04.069638 [ 41 ] {3894ca6e-0aa6-40ec-ac36-65359d2b95c5} <Information> RestoreCoordinationRemote: Waiting for the stage completed
[clickhouse-02] 2023.07.24 02:59:04.075955 [ 41 ] {3894ca6e-0aa6-40ec-ac36-65359d2b95c5} <Error> executeQuery: Code: 599. DB::Exception: Got error from clickhouse%2D03:9000. DB::Exception: Backup Disk('blob_storage_disk', 'test-21-07-23') not found. Stack trace:

0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000c2f4404 in /usr/bin/clickhouse
1. ? @ 0x00000000083f0150 in /usr/bin/clickhouse
2. DB::BackupImpl::readBackupMetadata() @ 0x00000000100bdd48 in /usr/bin/clickhouse
3. DB::BackupImpl::open(std::shared_ptr<DB::Context const> const&) @ 0x00000000100bad24 in /usr/bin/clickhouse
4. DB::BackupImpl::BackupImpl(String const&, DB::BackupImpl::ArchiveParams const&, std::optional<DB::BackupInfo> const&, std::shared_ptr<DB::IBackupReader>, std::shared_ptr<DB::Context const> const&) @ 0x00000000100ba7c0 in /usr/bin/clickhouse
5. ? @ 0x00000000100ce968 in /usr/bin/clickhouse
6. DB::BackupsWorker::doRestore(std::shared_ptr<DB::ASTBackupQuery> const&, String const&, String const&, DB::BackupInfo const&, DB::RestoreSettings, std::shared_ptr<DB::IRestoreCoordination>, std::shared_ptr<DB::Context>, bool) @ 0x000000001002549c in /usr/bin/clickhouse
7. ? @ 0x0000000010029e70 in /usr/bin/clickhouse
8. ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::worker(std::__list_iterator<ThreadFromGlobalPoolImpl<false>, void*>) @ 0x000000000c3b2f5c in /usr/bin/clickhouse
9. void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'(), void ()>>(std::__function::__policy_storage const*) @ 0x000000000c3b55dc in /usr/bin/clickhouse
10. ThreadPoolImpl<std::thread>::worker(std::__list_iterator<std::thread, void*>) @ 0x000000000c3af384 in /usr/bin/clickhouse
11. ? @ 0x000000000c3b4468 in /usr/bin/clickhouse
12. start_thread @ 0x0000000000007624 in /usr/lib/aarch64-linux-gnu/libpthread-2.31.so
13. ? @ 0x00000000000d149c in /usr/lib/aarch64-linux-gnu/libc-2.31.so
. (BACKUP_NOT_FOUND) (version 23.6.2.18 (official build)) (from 127.0.0.1:58166) (in query: RESTORE TABLE test AS test_restored ON CLUSTER cluster_1S_3R FROM Disk('blob_storage_disk', 'test-21-07-23')), Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000c2f4404 in /usr/bin/clickhouse
1. DB::readException(DB::ReadBuffer&, String const&, bool) @ 0x000000000c349dec in /usr/bin/clickhouse
2. DB::BackupCoordinationStageSync::readCurrentState(std::vector<String, std::allocator<String>> const&, std::vector<String, std::allocator<String>> const&, String const&) const @ 0x0000000010058990 in /usr/bin/clickhouse
3. DB::BackupCoordinationStageSync::waitImpl(std::vector<String, std::allocator<String>> const&, String const&, std::optional<std::chrono::duration<long long, std::ratio<1l, 1000l>>>) const @ 0x0000000010054804 in /usr/bin/clickhouse
4. DB::BackupsWorker::doRestore(std::shared_ptr<DB::ASTBackupQuery> const&, String const&, String const&, DB::BackupInfo const&, DB::RestoreSettings, std::shared_ptr<DB::IRestoreCoordination>, std::shared_ptr<DB::Context>, bool) @ 0x0000000010025dc0 in /usr/bin/clickhouse
5. DB::BackupsWorker::startRestoring(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context>) @ 0x000000001001f4c4 in /usr/bin/clickhouse
6. DB::BackupsWorker::start(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context>) @ 0x000000001001e0a8 in /usr/bin/clickhouse
7. DB::InterpreterBackupQuery::execute() @ 0x0000000010e16938 in /usr/bin/clickhouse
8. ? @ 0x0000000011179854 in /usr/bin/clickhouse
9. DB::executeQuery(String const&, std::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum) @ 0x00000000111764e4 in /usr/bin/clickhouse
10. DB::TCPHandler::runImpl() @ 0x0000000011d5ddd8 in /usr/bin/clickhouse
11. DB::TCPHandler::run() @ 0x0000000011d70ae4 in /usr/bin/clickhouse
12. Poco::Net::TCPServerConnection::start() @ 0x0000000012a275e4 in /usr/bin/clickhouse
13. Poco::Net::TCPServerDispatcher::run() @ 0x0000000012a28b00 in /usr/bin/clickhouse
14. Poco::PooledThread::run() @ 0x0000000012be98bc in /usr/bin/clickhouse
15. Poco::ThreadImpl::runnableEntry(void*) @ 0x0000000012be7184 in /usr/bin/clickhouse
16. start_thread @ 0x0000000000007624 in /usr/lib/aarch64-linux-gnu/libpthread-2.31.so
17. ? @ 0x00000000000d149c in /usr/lib/aarch64-linux-gnu/libc-2.31.so


0 rows in set. Elapsed: 0.248 sec.

Received exception from server (version 23.6.2):
Code: 599. DB::Exception: Received from localhost:9000. DB::Exception: Got error from clickhouse%2D03:9000. DB::Exception: Backup Disk('blob_storage_disk', 'test-21-07-23') not found. (BACKUP_NOT_FOUND)
```

**Comment by tavplubix** [comment]
Looks like a duplicate of https://github.com/ClickHouse/ClickHouse/issues/47729 and https://github.com/ClickHouse/ClickHouse/issues/52303

@vitlibar, please take a look

**Comment by vitlibar** [comment]
Azure blob storage is not fully supported as a destination for backups. The main problem is that like for any other disk type with a remote storage when ClickHouse writes something to an azure blob storage it also writes some metadata locally to `/var/lib/clickhouse/disks/blob_storage_disk/` and this metadata is absolutely necessary but it's not stored in the backup and it's not automatically replicated between hosts. So you can get into troubles with azure blob storage's backups even without the `on cluster` clause.

To solve the problem I can recommend using S3 API for Azure blob storage and use the command `BACKUP [ON CLUSTER...] ... TO S3(...)`

**Comment by vitlibar** [comment]
Or maybe later we can implement in ClickHouse a new command to use with Azure blob storage:

```
BACKUP [ON CLUSTER...] ... TO AzureBlobStorage(...)
```

**Comment by alifirat** [comment]
Hi @JeremyChen-179 

If you want to avoid data transfer between Azure and AWS, I've shared my experience on using s3proxy with Azure Blob Storage backend on ClickHouse

https://engineering.contentsquare.com/2023/clickhouse-on-azure/
